quantum/util/impl/quantum_sequencer_lite_impl.h

```diff
@@ -0,0 +1,475 @@
+    _exceptionCallback(configuration.getExceptionCallback()),
+    _taskStats(std::make_shared<SequenceKeyStatisticsWriter>())
+{
+}
+
+template <class SequenceKey, class Hash, class KeyEqual, class Allocator>
+bool
+SequencerLite<SequenceKey, Hash, KeyEqual, Allocator>::addPendingTask(
+    const std::shared_ptr<SequencerLiteTask<SequenceKey>>& task)
+{
+    bool empty = _universalTaskQueue._tasks.empty();
+    _universalTaskQueue._tasks.push_back(task);
+    if (not empty)
+    {
+        ++task->_pendingKeyCount;
+    }
+    _universalTaskQueue._stats->incrementPostedTaskCount();
+    _universalTaskQueue._stats->incrementPendingTaskCount();
+
+    return empty;
+}
+
+template <class SequenceKey, class Hash, class KeyEqual, class Allocator>
+bool
+SequencerLite<SequenceKey, Hash, KeyEqual, Allocator>::addPendingTask(
+    const SequenceKey& key,
+    const std::shared_ptr<SequencerLiteTask<SequenceKey>>& task)
+{
+    typename PendingTaskQueueMap::iterator it = _pendingTaskQueueMap.find(key);
+    if (it == _pendingTaskQueueMap.end())
+    {
+        it = _pendingTaskQueueMap.emplace(key, SequencerLiteKeyData<SequenceKey>()).first;
+        // add each universal task to this queue before the task
+        for(auto& universalTask : _universalTaskQueue._tasks)
+        {
+            it->second._tasks.push_back(universalTask);
+            ++universalTask->_pendingKeyCount;
+        }
+    }
+    bool empty = it->second._tasks.empty();
+    it->second._tasks.push_back(task);
+    if (not empty)
+    {
+        ++task->_pendingKeyCount;
+    }
+    if (not task->_universal)
+    {
+        task->_keyData.push_back(&it->second);
+        it->second._stats->incrementPostedTaskCount();
+        it->second._stats->incrementPendingTaskCount();
+    }
+    return empty;
+}
+
+template <class SequenceKey, class Hash, class KeyEqual, class Allocator>
+std::shared_ptr<SequencerLiteTask<SequenceKey>>
+SequencerLite<SequenceKey, Hash, KeyEqual, Allocator>::removePending(
+    SequencerLiteKeyData<SequenceKey>& entry,
+    const std::shared_ptr<SequencerLiteTask<SequenceKey>>& task)
+{
+    // delete this entry
+    if (entry._tasks.empty() or
+        entry._tasks.front() != task)
+    {
+        return std::shared_ptr<SequencerLiteTask<SequenceKey>>();
+    }
+
+    entry._tasks.pop_front();
+    if (entry._tasks.empty())
+    {
+        return std::shared_ptr<SequencerLiteTask<SequenceKey>>();
+    }
+    std::shared_ptr<SequencerLiteTask<SequenceKey>> next = entry._tasks.front();
+    // the new front has one less key to wait
+    return 0 == --next->_pendingKeyCount
+        ? next
+        : std::shared_ptr<SequencerLiteTask<SequenceKey>>();
+}
+
+template <class SequenceKey, class Hash, class KeyEqual, class Allocator>
+void
+SequencerLite<SequenceKey, Hash, KeyEqual, Allocator>::removePending(
+    const std::shared_ptr<SequencerLiteTask<SequenceKey>>& task)
+{
+    std::lock_guard<std::mutex> lock(_mutex);
```

Comment:
yes I was thinking how we can get rid of this std::mutex. I can't switch to quantum::Mutex yet because there's no context in SequencerLite::enqueue, and I need it to lock the mutex there for inserting the task to the queue. Notice that removePending is not very heavy and is actually a constant-time for non-universal tasks (provided the number of keys assigned to each task is a constant) . So locking the std::mutex here may not be a big deal. 

One way of switching to quantum::Mutex instead would be to enqueue each task to the control queue first and then in the control coroutine update the actual queues under the quantum::Mutex.